Improve Generic code path + Gentoo + tmpfs building

AdelKS opened the pull request:
Hello,

So everything is in the title, I did do some small fixes too. I tested on Ubuntu, Fedora and Gentoo. It all seems good!

Adel

LecrisUT commented:
I have confirmed the permission error. At a second installation, right after requesting sudo permission and performing the cleaning of the old linux-src-git, the root usage is maintained when pulling and preparing the next session.

AdelKS commented:
Thank you for your feedback. Actually, with the new approach of this PR, there should be no root owned files in `linux-src-git` any longer. I will fix that and ask you to confirm :D

LecrisUT commented:
Quick question, doesn't the makefile support `BUILDDIR` and such to make the cleanup and copying easier?

Also what's the purpose of removing the `*.ko` files after `modules_install` and before `make install`?
https://github.com/Frogging-Family/linux-tkg/blob/a5685a4086b377d4fe70e09b9cf9ed2510ab7340/install.sh#L497-L505
Are these properly ported over to `/lib/modules` during `modules_install` and no longer needed?

AdelKS commented:
> Quick question, doesn't the makefile support BUILDDIR and such to make the cleanup and copying easier?

I am not aware of this option, if you have the documentation at hand, please refer me to it. Otherwise I will look it up. Maybe it will enable me to write shorter and cleaner code.

> Are these properly ported over to /lib/modules during modules_install and no longer needed?

Yes I think so, my kernel properly boots after that. Therefore `modules_install` actually copies over the modules so they are no longer needed in the original sources, I believe.

I am not the most knowledgeable in these things and feedback is welcome. thanks!

LecrisUT commented:
Here's what I found about the output directory
https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/Documentation/admin-guide/README.rst#build-directory-for-the-kernel
I guess the following will work. I'll test it on my own as well
```shell
make O=./build
```
Using `make config` or equivalents we should find the other configurable options as well.

LecrisUT commented:
This part is confusing
https://github.com/torvalds/linux/blob/27151f177827d478508e756c7657273261aaf8a9/Makefile#L1331-L1336
From how I read it, `make install` should automatically call  dkms and install the available modules (probably need to unset `INSTALL_PATH`). I'll double check how the nvidia driver and stuff are installed and if I can manually add them via dkms to the new kernels.

LecrisUT commented:
A very specific issue for `akmods` needed for nvidia driver
```
akmodsbuild: error: line 31: Invalid version (double separator '-'): 5.14.1-tkg-cfs: Provides:         kernel-modules-for-kernel = 5.14.1-tkg-cfs
```
Seems the naming convention has to be considered

LecrisUT commented on install.sh:
```diff
@@ -370,80 +380,155 @@ if [ "$1" = "install" ]; then
-      else
-        _kernelname=$_basekernel.${_kernel_subver}-$_kernel_flavor
+    msg2 "Building kernel"
+    _timed_build make ${llvm_opt} -j ${_thread_num}
```
Can use `make O=./build` here, but you also have to add it here
https://github.com/AdelKS/linux-tkg/blob/a5685a4086b377d4fe70e09b9cf9ed2510ab7340/linux-tkg-config/prepare#L1303
and also fix the location of where the `.config` file would be found.

I have not tested if it's needed in other parts after, hopefully it is enough to find them all.

LecrisUT replied:
Honestly, it would be so awesome if the kernel had a `cmake` compatibility, where all of those standards are baked in. Also `ninja` builds for faster compilation.

AdelKS replied:
Well, the kernel is super old and developed its own [build system](https://www.kernel.org/doc/html/latest/kbuild/index.html). Seeing that change is very unlikely.

AdelKS commented:
> My recommendation based on the other formats is [...]

Yea that Fedora quirk is already done in the `Fedora` install path. This is "Generic" and does not have to follow those limitations as other distros do not have that issue. But given, that the call `make rpm-pkg` produces packages that replaces the currently installed kernel, I am thinking of making the Fedora install script use `Generic` and only mod it with this naming quirk for example.

AdelKS commented on install.sh:
```diff
@@ -91,26 +91,26 @@ _linux_git_branch_checkout() {
-    if [ "${_distro}" = "Generic" ]; then
+    if [[ "${_distro}" =~ ^(Generic|Gentoo)$ ]]; then
       msg2 "Please enter your sudo password so the script can clean root owned files from the kernel sources"
+      sudo git reset --hard HEAD
```
Yes I will do that. All sudo calls on the folder `linux-src-git` will be removed as the kernel will not be installed from there any longer.

LecrisUT commented:
> But given, that the call `make rpm-pkg` produces packages that replaces the currently installed kernel

Maybe there is a better solution down the line to not have it delete the old package.

> This is "Generic" and does not have to follow those limitations as other distros do not have that issue

Hmm, would it create issues to other distros though? Does Ubuntu and the likes not follow the same convention? I've caught this issue from the `akmods` usage, so maybe that is the standard that all should use?

LecrisUT commented on install.sh:
```diff
@@ -370,80 +380,155 @@ if [ "$1" = "install" ]; then
+    echo "    sudo grub-mkconfig -o /boot/grub/grub.cfg"
+
+    msg2 "Note: Uninstalling requires manual intervention, use './install.sh uninstall-help' for more information."
+    read -p "Continue ? Y/[n]: " _continue
```
I don't know why, but sometimes the installer just refuses to continue past this point even as I enter `Y`. e.g.:
```
 -> Note: Uninstalling requires manual intervention, use './install.sh uninstall-help' for more information.
Continue ? Y/[n]: Y
 -> exit cleanup done
```
Maybe there's a better failsafe version of the prompt that can trim spaces, invisible characters, etc.

AdelKS replied:
Can you print the value of `_continue` in that `if` block, with quotes (to see white spaces), to see what's wrong ?

AdelKS commented:
> Maybe there is a better solution down the line to not have it delete the old package.

Yes, to have the package name changed, `make rpm-pkg` down the line calls `makerpm` or whatever the tool that makes PRMs, and maybe there's a way to give an option to make it overwrite the package name in the spec file. I looked it up for a while then gave up. If you have knowledge in RPM packaging and would like to help there, it would be awesome!

> Hmm, would it create issues to other distros though? Does Ubuntu and the likes not follow the same convention?

The convention is the same except for dashes being replaced by underscores quirk. Dashes are actually broadly used in deb based distros (examples of kernel versions are in #296 like `linux-image-5.11.0-31-generic`) and Arch.

> I've caught this issue from the akmods usage, so maybe that is the standard that all should use?

Fedora afaik is the only distro using akmod ? Arch and Ubuntu use dkms, gentoo it's the package manager that handles module rebuilding.

> Probably the parsing error is because it thinks it's looking at a revision number, so we should at least deliminate with a period at that point.

We just need to not use dashes in Fedora and Suse x)

LecrisUT commented:
Update on trying to install the nvidia driver, it seems that the `akmods` are even more demanding:
```log
2021/09/09 01:00:40 akmods: Building RPM using the command '/sbin/akmodsbuild --kernels 5.14.1.tkg-pds /usr/src/akmods/nvidia-kmod.latest'
warning: user mockbuild does not exist - using root
warning: group mock does not exist - using root
warning: user mockbuild does not exist - using root
warning: group mock does not exist - using root
error: Package has no %description: kmod-nvidia-5.14.1.tkg-pds
Installing /usr/src/akmods/nvidia-kmod.latest
Building target platforms: x86_64
Building for target x86_64
2021/09/09 01:00:40 akmodsbuild: * Rebuilding /usr/src/akmods/nvidia-kmod.latest for kernel(s) 5.14.1.tkg-pds: warning: user mockbuild does not exist - using root
2021/09/09 01:00:40 akmodsbuild: warning: group mock does not exist - using root
2021/09/09 01:00:40 akmodsbuild: warning: user mockbuild does not exist - using root
2021/09/09 01:00:40 akmodsbuild: warning: group mock does not exist - using root
2021/09/09 01:00:40 akmodsbuild: error: Package has no %description: kmod-nvidia-5.14.1.tkg-pds
2021/09/09 01:00:40 akmodsbuild: Installing /usr/src/akmods/nvidia-kmod.latest
2021/09/09 01:00:40 akmodsbuild: Building target platforms: x86_64
2021/09/09 01:00:40 akmodsbuild: Building for target x86_64
2021/09/09 01:00:40 akmodsbuild: 
2021/09/09 01:00:40 akmods: Building rpms failed; see /var/cache/akmods/nvidia/470.63.01-1-for-5.14.1.tkg-pds.failed.log for details
```

AdelKS commented:
> Update on trying to install the nvidia driver, it seems that the akmods are even more demanding

It works with the normal approach when using `Fedora` in the script ?

LecrisUT commented:
I think what the error is referring to is that the kernel is not installed via `rpm` so it does not have metadata such as `%description` to work on.

AdelKS commented:
I see, so the `Generic` approach for Fedora is out of question. The only hope left is to be able to change the package name. I looked up how to change it _after_ the RPMs are created, but I didn't find any good documentation. Because the actual files that are installed do not conflict with any other kernel. It's just in the spec that it says "this package is 'kernel' and upgrade/downgrades kernel"

LecrisUT commented on install.sh:
```diff
@@ -370,80 +372,165 @@ if [ "$1" = "install" ]; then
+    msg2 "Creating initramfs"
+    sudo dracut --force --hostonly ${_dracut_options} --kver $_kernelname
+    msg2 "Updating GRUB"
+    sudo grub-mkconfig -o /boot/grub/grub.cfg
```
Some distros use `grub2-mkconfig` and similar naming conventions. You also might want to install to `/boot/efi/...` instead. Not sure why the script managed to run on Fedora though.

AdelKS replied:
Well, in any case, the `Generic` approach should not be used by rpm users. Hopefully the other approach of renaming the package names works so this is not needed.

AdelKS commented:
Renaming the kernel package name in Fedora is not useful as kernels can still be installed in parallel, up to a certain number of installed kernels. So everything is good. I dropped those changes and moved them to [this branch](https://github.com/AdelKS/linux-tkg/tree/fedora-kernel-rename) if ever it's needed in the future.

AdelKS commented:
Alright, I think we are all good to go @Tk-Glitch

![One merge please -frog v2](https://user-images.githubusercontent.com/13605217/133891261-bf606832-d18f-479c-a44e-a151503d5882.png)

Tk-Glitch commented:
Thanks guys! :frog: :heart:
